source-academy/cadet-frontend
# Contests dev

### Description



**Summary of changes** 
Contest Voting: 
- Support for contest voting assessments for students to vote for contest entries. 
- Implemented through a 'practical' assessment with contest voting question and students can view entries and vote
via the contest votes tab. 
- Added front end layer of validation of input to ensure that all scores given are unique and all scores are within the range [1, # of contest entries to vote assigned].  

Leaderboard: 
- display of leaderboard entries. 
- the leaderboard is rolling and updates every 4 hours for ongoing contest votes and updates before everything is done. 
- the rolling leaderboard is updated every 2 hours(only for ongoing contest votes) and 1 final time at midnight when the contest ends. 
( future update to score the penalty after contest is submitted?) -> this will remove the need for a scheduler. 

### Type of change


### How to test



I have included snapshot testing for the SideContentContestVotingContainer as well as testing for the Contest Voting question
rendering correctly for submitted and yet-to-be submitted submissions. 

I have also included several behavior testing including: 
- testing that the number of entries is displayed dynamically based on the number of entries passed.
- testing the validation logic for first layer validation of contest vote submissions. 
- testing that voting is disabled after submission for students 
- testing that the editor updates when clicking on individual entries. 

To test manually,
XML setup: 
Set up the linked contest + voters in the xml. The rest of the xml file config is similar to other assessments. 
![image](https://user-images.githubusercontent.com/55353265/117435228-353f6380-af60-11eb-9f67-a06989585d51.png)
  
Staff/admin uploading of contest voting assessment (the votes are automatically assigned to each user indicated in the xml file once the assessment is uploaded, hence, please only upload after the contest ends.) 
![image](https://user-images.githubusercontent.com/55353265/117435143-1a6cef00-af60-11eb-963f-3a69230203a5.png)

First, we need the corresponding backend changes in cadet.
![image](https://user-images.githubusercontent.com/55353265/112585444-c76a2d00-8e34-11eb-9295-7263e55d137d.png)
Clicking each entry changes the program on the code editor workspace. 

The input fields are disabled when the contest voting submission has been submitted. 
![image](https://user-images.githubusercontent.com/55353265/112585348-9984e880-8e34-11eb-8512-6d520ecbebb7.png)

Viewing of rolling leaderboard: 
![image](https://user-images.githubusercontent.com/55353265/117434942-dc6fcb00-af5f-11eb-8fc9-0fdb4f29cf25.png)


### Checklist




I have left some documentation regarding the purpose of the container - to separate rendering from API calls etc within the components as necessary. 

Known issues: 
The contest voting component does not support mobile as of now.
